File: src/main/java/com/couchbase/cblite/CBLRevision.java
package com.couchbase.cblite;

import com.couchbase.cblite.internal.InterfaceAudience;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Stores information about a revision -- its docID, revID, and whether it's deleted.
 *
 * It can also store the sequence number and document contents (they can be added after creation).
 */
public abstract class CBLRevision {


    /**
     * The sequence number of this revision.
     */
    protected long sequence;

    /**
     * The revisions's owning database.
     */
    protected CBLDatabase database;

    /**
     * The document this is a revision of
     */
    protected CBLDocument document;

    /**
     * Constructor
     */
    @InterfaceAudience.Private
    CBLRevision() {
        super();
    }

    /**
     * Constructor
     */
    @InterfaceAudience.Private
    protected CBLRevision(CBLDocument document) {
        this.document = document;
    }

    /**
     * Get the revision's owning database.
     */
    @InterfaceAudience.Public
    public CBLDatabase getDatabase() {
        return database;
    }

    /**
     * Get the document this is a revision of.
     */
    @InterfaceAudience.Public
    public CBLDocument getDocument() {
        return document;
    }

    /**
     * Gets the Revision's id.
     */

    @InterfaceAudience.Public
    public abstract String getId();


    /**
     * Does this revision mark the deletion of its document?
     * (In other words, does it have a "_deleted" property?)
     */
    @InterfaceAudience.Public
    boolean isDeletion() {
        Object deleted = getProperty("_deleted");
        if (deleted == null) {
            return false;
        }
        Boolean deletedBool = (Boolean) deleted;
        return deletedBool.booleanValue();
    }

    /**
     * The contents of this revision of the document.
     * Any keys in the dictionary that begin with "_", such as "_id" and "_rev", contain CouchbaseLite metadata.
     *
     * @return contents of this revision of the document.
     */
    @InterfaceAudience.Public
    public abstract Map<String,Object> getProperties();


    /**
     * The user-defined properties, without the ones reserved by CouchDB.
     * This is based on -properties, with every key whose name starts with "_" removed.
     *
     * @return user-defined properties, without the ones reserved by CouchDB.
     */
    @InterfaceAudience.Public
    public Map<String,Object> getUserProperties() {

        Map<String,Object> result = new HashMap<String, Object>();

        Map<String,Object> sourceMap = getProperties();
        for (String key : sourceMap.keySet()) {
            if (!key.startsWith("_")) {
                result.put(key, sourceMap.get(key));
            }
        }
        return Collections.unmodifiableMap(result);
    }

    /**
     * The names of all attachments
     * @return
     */
    @InterfaceAudience.Public
    public List<String> getAttachmentNames() {
        Map<String, Object> attachmentMetadata = getAttachmentMetadata();
        ArrayList<String> result = new ArrayList<String>();
        result.addAll(attachmentMetadata.keySet());
        return result;
    }

    /**
     * All attachments, as CBLAttachment objects.
     */
    @InterfaceAudience.Public
    public List<CBLAttachment> getAttachments() {
        List<CBLAttachment> result = new ArrayList<CBLAttachment>();
        List<String> attachmentNames = getAttachmentNames();
        for (String attachmentName : attachmentNames) {
            result.add(getAttachment(attachmentName));
        }
        return result;
    }

    /**
     * Shorthand for getProperties().get(key)
     */
    @InterfaceAudience.Public
    public Object getProperty(String key) {
        return getProperties().get(key);
    }

    /**
     * Looks up the attachment with the given name (without fetching its contents yet).
     */
    @InterfaceAudience.Public
    public CBLAttachment getAttachment(String name) {
        Map<String, Object> attachmentMetadata = getAttachmentMetadata();
        if (attachmentMetadata == null) {
            return null;
        }
        return new CBLAttachment(this, name, attachmentMetadata);
    }



    Map<String, Object> getAttachmentMetadata() {
        return (Map<String, Object>) getProperty("_attachments");
    }

    @Override
    public boolean equals(Object o) {
        boolean result = false;
        if(o instanceof CBLSavedRevision) {
            CBLSavedRevision other = (CBLSavedRevision)o;
            if(document.getId().equals(other.getDocument().getId()) && getId().equals(other.getId())) {
                result = true;
            }
        }
        return result;
    }

    @Override
    public int hashCode() {
        return document.getId().hashCode() ^ getId().hashCode();
    }


    void setSequence(long sequence) {
        this.sequence = sequence;
    }

    long getSequence() {
        return sequence;
    }

    @Override
    public String toString() {
        return "{" + this.document.getId() + " #" + this.getId() + (isDeletion() ? "DEL" : "") + "}";
    }

    /**
     * Generation number: 1 for a new document, 2 for the 2nd revision, ...
     * Extracted from the numeric prefix of the revID.
     */
    int getGeneration() {
        return generationFromRevID(getId());
    }

    static int generationFromRevID(String revID) {
        int generation = 0;
        int dashPos = revID.indexOf("-");
        if(dashPos > 0) {
            generation = Integer.parseInt(revID.substring(0, dashPos));
        }
        return generation;
    }

}

File: Couchbase.Lite/Couchbase.Lite/CBLRevision.cs
/**
 * Couchbase Lite for .NET
 *
 * Original iOS version by Jens Alfke
 * Android Port by Marty Schoch, Traun Leyden
 * C# Port by Zack Gramana
 *
 * Copyright (c) 2012, 2013 Couchbase, Inc. All rights reserved.
 * Portions (c) 2013 Xamarin, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

using System;
using System.Collections.Generic;
using Couchbase;
using Sharpen;

namespace Couchbase
{
	/// <summary>Stores information about a revision -- its docID, revID, and whether it's deleted.
	/// 	</summary>
	/// <remarks>
	/// Stores information about a revision -- its docID, revID, and whether it's deleted.
	/// It can also store the sequence number and document contents (they can be added after creation).
	/// </remarks>
	public class CBLRevision
	{
		private string docId;

		private string revId;

		private bool deleted;

		private CBLBody body;

		private long sequence;

		private CBLDatabase database;

		public CBLRevision(string docId, string revId, bool deleted, CBLDatabase database
			)
		{
			this.docId = docId;
			this.revId = revId;
			this.deleted = deleted;
			this.database = database;
		}

		public CBLRevision(CBLBody body, CBLDatabase database) : this((string)body.GetPropertyForKey
			("_id"), (string)body.GetPropertyForKey("_rev"), (((bool)body.GetPropertyForKey(
			"_deleted") != null) && ((bool)body.GetPropertyForKey("_deleted") == true)), database
			)
		{
			this.body = body;
		}

		public CBLRevision(IDictionary<string, object> properties, CBLDatabase database) : 
			this(new CBLBody(properties), database)
		{
		}

		public virtual IDictionary<string, object> GetProperties()
		{
			IDictionary<string, object> result = null;
			if (body != null)
			{
				result = body.GetProperties();
			}
			return result;
		}

		public virtual void SetProperties(IDictionary<string, object> properties)
		{
			this.body = new CBLBody(properties);
			// this is a much more simplified version that what happens on the iOS.  it was
			// done this way due to time constraints, so at some point this needs to be
			// revisited to port the remaining functionality.
			IDictionary<string, object> attachments = (IDictionary<string, object>)properties
				.Get("_attachments");
			if (attachments != null && attachments.Count > 0)
			{
				foreach (string attachmentName in attachments.Keys)
				{
					IDictionary<string, object> attachment = (IDictionary<string, object>)attachments
						.Get(attachmentName);
					// if there is actual data in this attachment, no need to try to install it
					if (attachment.ContainsKey("data"))
					{
						continue;
					}
					CBLStatus status = database.InstallPendingAttachment(attachment);
					if (status.IsSuccessful() == false)
					{
						string msg = string.Format("Unable to install pending attachment: %s.  Status: %d"
							, attachment.ToString(), status.GetCode());
						throw new InvalidOperationException(msg);
					}
				}
			}
		}

		public virtual byte[] GetJson()
		{
			byte[] result = null;
			if (body != null)
			{
				result = body.GetJson();
			}
			return result;
		}

		public virtual void SetJson(byte[] json)
		{
			this.body = new CBLBody(json);
		}

		public override bool Equals(object o)
		{
			bool result = false;
			if (o is Couchbase.CBLRevision)
			{
				Couchbase.CBLRevision other = (Couchbase.CBLRevision)o;
				if (docId.Equals(other.docId) && revId.Equals(other.revId))
				{
					result = true;
				}
			}
			return result;
		}

		public override int GetHashCode()
		{
			return docId.GetHashCode() ^ revId.GetHashCode();
		}

		public virtual string GetDocId()
		{
			return docId;
		}

		public virtual void SetDocId(string docId)
		{
			this.docId = docId;
		}

		public virtual string GetRevId()
		{
			return revId;
		}

		public virtual void SetRevId(string revId)
		{
			this.revId = revId;
		}

		public virtual bool IsDeleted()
		{
			return deleted;
		}

		public virtual void SetDeleted(bool deleted)
		{
			this.deleted = deleted;
		}

		public virtual CBLBody GetBody()
		{
			return body;
		}

		public virtual void SetBody(CBLBody body)
		{
			this.body = body;
		}

		public virtual Couchbase.CBLRevision CopyWithDocID(string docId, string revId)
		{
			System.Diagnostics.Debug.Assert(((docId != null) && (revId != null)));
			System.Diagnostics.Debug.Assert(((this.docId == null) || (this.docId.Equals(docId
				))));
			Couchbase.CBLRevision result = new Couchbase.CBLRevision(docId, revId, deleted, database
				);
			IDictionary<string, object> properties = GetProperties();
			if (properties == null)
			{
				properties = new Dictionary<string, object>();
			}
			properties.Put("_id", docId);
			properties.Put("_rev", revId);
			result.SetProperties(properties);
			return result;
		}

		public virtual void SetSequence(long sequence)
		{
			this.sequence = sequence;
		}

		public virtual long GetSequence()
		{
			return sequence;
		}

		public override string ToString()
		{
			return "{" + this.docId + " #" + this.revId + (deleted ? "DEL" : string.Empty) + 
				"}";
		}

		/// <summary>Generation number: 1 for a new document, 2 for the 2nd revision, ...</summary>
		/// <remarks>
		/// Generation number: 1 for a new document, 2 for the 2nd revision, ...
		/// Extracted from the numeric prefix of the revID.
		/// </remarks>
		public virtual int GetGeneration()
		{
			return GenerationFromRevID(revId);
		}

		public static int GenerationFromRevID(string revID)
		{
			int generation = 0;
			int dashPos = revID.IndexOf("-");
			if (dashPos > 0)
			{
				generation = System.Convert.ToInt32(Sharpen.Runtime.Substring(revID, 0, dashPos));
			}
			return generation;
		}
	}
}
